### PR TITLE
Issue #126 hybrid dropout bug

### LINKED ISSUE #126: [BUG] Incorrect factor formula in hybrid dropout
### Describe the bug

The factor applied at the end of hybrid dropout is currently:

```
            probability = zeros / total_neurons
```

However, equation 10 in the [paper](https://www.sciencedirect.com/science/article/pii/S1566253524000770) would translate to:

```
            probability = non_zeros / total_neurons
```

But I have come to realize that the equation in the paper is also incorrect. With a `p` of 0.5 and a `beta` of 0, both of these equations would mean that the factor was roughly 0.5. So we are dropping half the weights and instead of compensating for the scaling by multiplying by 2 (as is indicated in the pytorch [dropout docs](https://docs.pytorch.org/docs/stable/generated/torch.nn.Dropout.html) ), we are exacerbating the scaling issue.

Therefore, I propose that the correct line of code here would be:

```
            probability = total_neurons / (non_zeros + 1e-9)
```

This is the multiplicative inverse of the equation from the paper and it is in line with how this is done in normal dropout. Empirically, using either the current hybrid dropout factor or the one from the paper yields very poor results and makes validation loss always higher than training loss because on average, there is an p^2 scaling (the dropout and factor both tend to cause a scaling of around factor p) applied in training but not in validation. While using the multiplicative inverse of the scaling factor from the paper yields a much more normal training setup with seemingly a bit of improvement in loss.

### Steps/Code to reproduce the bug

_No response_

### Expected results

Since dropout prevents a model from using all neurons at train time and allows them at validation time, the expectation is that validation loss should be lower than training loss until and unless some other factor (such as overfitting) comes into play.

### Actual results

Validation loss is always higher than training loss since the factor calculation does not mitigate the scaling difference caused by dropout during training.

### PR DESCRIPTION
**Pull request by franberchez**
Bug fixed in files `hybrid_dropout.py` and `test_hybrid_dropout.py` regarding the issue #126  

- Fixed incorrect scaling factor in HybridDropout (now uses 1 / keep_prob) and removed unused `zeros` variable.
- Added unit test to verify proper dropout scaling behavior.

### CODE REVIEW

**Review by victormvy — APPROVED** [review]
Looks good to me! Thanks!
